src/Tokenomia/Vesting/PrivateSale.hs

```diff
@@ -0,0 +1,327 @@
+  let invTxhs = privateSaleToTxhs ps in
+    let txhs = verifyTxHashList invTxhs treasAddrTxs in
+      if length txhs == length invTxhs then do
+        verifyTxs (getPsInvestments ps) txhs
+        else throwError . BlockFrostError . B.BlockfrostError $ "Unequal lengths, means missing TXs"
+
+jsonToPrivateSale ::
+  (MonadIO m
+  , MonadError  TokenomiaError m
+  , MonadReader Environment m)
+  => FilePath 
+  -> m PrivateSale
+jsonToPrivateSale jsonFilePath = do
+    fileContents <- liftIO . readFile $ jsonFilePath
+    case (decode fileContents :: Maybe PrivateSale) of
+        Nothing -> throwError . BlockFrostError . B.BlockfrostError $ "Unable to parse JSON"
+        Just ps -> return ps
+
+privateSaleToTxhs :: PrivateSale -> [B.TxHash]
+privateSaleToTxhs ps = invTxhs 
+
+  where
+    invTxhs :: [B.TxHash]
+    invTxhs = (^. invTx) <$> invs
+
+    invs :: [Investment]
+    invs = getPsInvestments ps
+
+verifyTxs ::
+  (MonadIO m
+  , MonadError  TokenomiaError m
+  , MonadReader Environment m)
+  => [Investment]
+  -> [B.TxHash]
+  -> m ()
+verifyTxs invs = 
+    foldl (\z txh -> do
+    bfTx <- lift . getTxByTxHash $ txh                                      -- :: m B.Transaction
+    -- inv  <- MaybeT . return $ getInvByTxHash txh invs                    -- :: Maybe Investment
+    -- inv <- getInvByTxHash txh invs
+    let inv = getInvByTxHash txh invs in
+      let invTokenName = getTokenName $ inv ^. invAssetClass                  -- :: Text
+          bfAmts = bfTx ^. B.outputAmount                                     -- :: [B.Amount]
+          invAmount' = inv ^. invAmount in                                    -- :: Integer
+              let amt = getAmountByTokenName invTokenName bfAmts in           -- :: B.Amount
+              if getDiscreteAmount amt == invAmount' then z 
+              else throwError . BlockFrostError . B.BlockfrostError $ "Amts don't match"
+              ) (return ())
+        
+getTreasAddrTxs ::
+  (MonadIO m
+  , MonadError  TokenomiaError m
+  , MonadReader Environment m)
+  => PrivateSale 
+  -> m [B.AddressTransaction]
+getTreasAddrTxs ps = do
+  prj <- projectFromEnv''
+  liftIO $ B.runBlockfrost prj $ do
+    B.getAddressTransactions (ps ^. psAddress)
+  >>= (\case 
+      Left e    -> throwError $ BlockFrostError e
+      Right res -> return res)
+
+getTxByTxHash ::
+  (MonadIO m
+  , MonadError  TokenomiaError m
+  , MonadReader Environment m)
+  => B.TxHash
+  -> m B.Transaction
+getTxByTxHash txh = do
+  prj <- projectFromEnv''
+  liftIO $ B.runBlockfrost prj $ do
+    B.getTx txh
+  >>= (\case
+        Left e    -> throwError $ BlockFrostError e
+        Right res -> return res)
+
+getTokenName :: AssetClass -> Text
+getTokenName = pack . unpackChars . fromBuiltin . unTokenName . snd . unAssetClass
+        
+--TODO: use nonEmpty so I don't have to have a case for an empty list. then I don't need to return maybe
+--   and then I wont need to use MaybeT
```

Comment:
I sent you a message on Slack about this, but you can just use something like

```haskell
inv <- maybe (throwError . someFunctionReturningATokenomiaError) pure $ getInvByTxHash ...
```

with a signature of `B.TxHash -> [Investment] -> Investment`

Better yet: return `Either TokenomiaError Investment` and then use `liftEither` from `Control.Monad.Class.Error`